py3status/modules/usbguard.py

```diff
@@ -200,182 +82,94 @@ def post_config_hook(self):
-                "/org/usbguard/Devices",
                 "org.usbguard.Devices",
+                signal,
                 None,
+                None,
+                0,
+                lambda *args: self.py3.update(),
             )
-            for signal in ["DevicePolicyChanged", "DevicePresenceChanged"]:
-                self.bus.signal_subscribe(
-                    None, "org.usbguard.Devices", signal, None, None, 0, self.py3.update
-                )
-        except Exception:
-            raise Exception(STRING_USBGUARD_DBUS)
-
-        self.thread = Thread(target=self._start_loop)
-        self.thread.daemon = True
-        # self.thread.start()
 
-    def _start_loop(self):
-        pass
+        thread = Thread(target=lambda: GLib.MainLoop().run())
+        thread.daemon = True
+        thread.start()
 
-    def _get_devices(self, device_filter):
-        if device_filter in ["any", "None"]:
-            device_filter = "match"
-
-        devices = self.proxy.listDevices("(s)", device_filter)
-        new_devices = []
-
-        for device_id, device_string in devices:
-            try:
-                device = self.cache_devices[device_string]
-            except KeyError:
-                device = {"id": device_id}
-                for name, regex in self.device_keys:
-                    value = regex.findall(device_string) or None
-                    if value:
-                        value = (
-                            value[0]
-                            .encode("latin-1")
-                            .decode("unicode_escape")
-                            .encode("latin-1")
-                            .decode()
-                        )
-                    device[name] = value
-                self.cache_devices[device_string] = device
-            new_devices.append(device)
-
-        return deepcopy(new_devices)
+    def _get_devices(self):
+        try:
+            raw_devices = self.proxy.listDevices("(s)", "match")
+        except Exception:
+            raise Exception(STRING_USBGUARD_DBUS)
 
-    def _manipulate_devices(self, devices, device_filter):
-        new_device = []
-        format_device_list = None
-        self.device_length = len(devices)
-        for index, device in enumerate(devices, 1):
-            if device_filter == "any":
-                pass
-            elif device_filter == "None":
-                device["policy"] = None
-            elif device["policy"] != device_filter:
-                continue
-            for x in self.init["format_button"]:
-                composite = self.py3.safe_format(getattr(self, x), device)
-                device[x] = self.py3.composite_update(
-                    composite, {"index": "{}/{}".format(device["id"], x.split("_")[-1])}
+        devices = []
+        for device_id, string in raw_devices:
+            device = {"id": device_id}
+            for name, regex in self.keys:
+                value = regex.findall(string) or None
```

Comment:
I wanted to try `re.compile(a single long string)` instead of a list of keys to `re.compile` and maybe `encode/decode` the string first too.

Reply:
I'd like to avoid a single huge regex, because it will be difficult to understand it. Encode/decode string beforehand is nice idea, done.

Reply:
Long regex string should look like this.. and is similar to list.
```python
regex_string = (
    'first_part (123312)'
    'second_part (123312)'
    'third_part (123312)'
    'fourth_part (123312)'
    'fifth_part (123312)'
)
```

Reply:
But downsides are that this begins to depend on the order of keys in the string we search, and also we need to carefully treat the separators between the strings (e.g. `.*`, but maybe something more special for multiline strings)... Not sure the speed difference is worth it tbh.

Reply:
Well, we don't plug/unplug usb devices often so we can live with this as-is. :+1: